Fix Integration tests skipped with Redis cluster

maxbeckers opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #49285
| License       | MIT
| Doc PR        | n/a

It seems to me, that the `getenv('MESSENGER_REDIS_SENTINEL_MASTER')` in the `setUp()` might return `false` because it's unset. To ignore the sentinel, it must be null:
https://github.com/symfony/symfony/blob/090033332fa4eb5e7475111136b1cce260edffa4/src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php#L95-L100

juagarc4 committed:
[MonologBridge] FirePHPHandler::onKernelResponse throws PHP 8.1 deprecation when no user agent is set

xabbuh committed:
do not drop embed label classes

AdamKatzDev committed:
[Messenger][Cache] fixed CallbackInterface support in async expiration handler

Commit:
[Contracts] Fix setting $container before calling parent::setContainer in ServiceSubscriberTrait

chalasr committed:
[SecurityBundle] Fix `Security::login()` on specific firewall

maxbeckers commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -154,7 +154,7 @@ private static function initializeRedis(\Redis|Relay $redis, string $host, int $
     private static function initializeRedisCluster(?\RedisCluster $redis, array $hosts, string|array|null $auth, array $params): \RedisCluster
     {
-        $redis ??= new \RedisCluster(null, $hosts, $params['timeout'], $params['read_timeout'], (bool) $params['persistent'], $auth, ...\defined('Redis::SCAN_PREFIX') ? [$params['ssl'] ?? null] : []);
+        $redis ??= new \RedisCluster(null, $hosts, $params['timeout'], $params['read_timeout'], isset($params['persistent_id']), $auth, ...\defined('Redis::SCAN_PREFIX') ? [$params['ssl'] ?? null] : []);
```
Here I changed it to `isset($params['persistent_id'])` because `persistent` is not a default option and this option seems to be for me for this case. Otherwise I'd have do add the default option for `persistent`.

nicolas-grekas replied:
`$params['persistent'] ?? false`?

maxbeckers replied:
fine with that one.

maxbeckers commented on .github/workflows/integration-tests.yml:
```diff
@@ -49,7 +49,7 @@ jobs:
           - 16379:6379
       redis-cluster:
-        image: grokzen/redis-cluster:latest
+        image: grokzen/redis-cluster:6.0.0
```
I guess using `latest` is not always the best way, because you don't know with which version the tests were running. In this image the `latest` tag should be `7.0.7` for now, but it doesn't match that version. And we are using redis 6.0.0 for all of our redis tests, so we should do it here this way as well.

Edit: The `latest` tag was added in #48761

stof replied:
We should probably use 6.2.8 instead of 6.0.0. According to https://relay.so/docs/1.x/installation#requirements, Relay requires Redis 6.2.7+

maxbeckers replied:
Then we should update redis and redis-cluster to this version IMHO. Maybe this fixes the skipped relay tests, I'll try this tomorrow.

maxbeckers replied:
I updated to 6.2.8. But the skipped tests are still failing.

maxbeckers commented:
For now there are failing 5 tests with `Invalid HELLO reply, Redis Server version may be too old (RELAY_ERR_COMPAT; commands.c:6660)`. I'll need to investigate a bit more of time to that or maybe someone else has an idea.

maxbeckers commented on src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/RelayExtIntegrationTest.php:
```diff
@@ -21,8 +21,45 @@
+
+    public function testConnectionSendAndGetDelayed()
+    {
+        self::markTestSkipped(self::SKIP_TEST_NOT_RUNNING_WITH_RELAY);
```
For now I added a skip to the tests not working with relay. Could be another issue to fix this tests working with relay. In this PR i focussed on the topic of the ticket, that cluster tests were not running. This problems came up after enabling this.

Seb33300 committed:
[Form] Skip password hashing on empty password

nicolas-grekas committed:
bug #49441 [Contracts] Fix setting $container before calling parent::setContainer in ServiceSubscriberTrait (edsrzf)

This PR was squashed before being merged into the 5.4 branch.

Discussion
----------

[Contracts] Fix setting $container before calling parent::setContainer in ServiceSubscriberTrait

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #49382
| License       | MIT

This changes the `setContainer` method of `ServiceSubscriberTrait` so that it calls `parent::setContainer` first, before updating `$this->container`. This is so that the parent method can work correctly if it relies on the old container value.

This is particularly relevant for `AbstractController`, since it returns the old container value. When used with `ServiceSubscriberTrait`, it returned the _new_ container rather than the old one. This, in turn, caused incorrect behavior in framework bundle's `ControllerResolver`, as described in #49382.

Commits
-------

bccb0747c7 [Contracts] Fix setting $container before calling parent::setContainer in ServiceSubscriberTrait

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/RelayExtIntegrationTest.php:
```diff
@@ -21,8 +21,45 @@
  */
 class RelayExtIntegrationTest extends RedisExtIntegrationTest
 {
+    public const SKIP_TEST_NOT_RUNNING_WITH_RELAY = 'This test doesn\'t work with relay.';
```
no need for the const, let's inline

maxbeckers replied:
Ok

nicolas-grekas commented:
I suppose we should merge this on 5.4?

maxbeckers commented:
@nicolas-grekas yes, i didn't check this, in the issue was mentioned the 6.3 branch. Is it possible to handle this during the merge or should I rebase it?

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -154,7 +154,7 @@ private static function initializeRedis(\Redis|Relay $redis, string $host, int $
     private static function initializeRedisCluster(?\RedisCluster $redis, array $hosts, string|array|null $auth, array $params): \RedisCluster
     {
-        $redis ??= new \RedisCluster(null, $hosts, $params['timeout'], $params['read_timeout'], (bool) $params['persistent'], $auth, ...\defined('Redis::SCAN_PREFIX') ? [$params['ssl'] ?? null] : []);
+        $redis ??= new \RedisCluster(null, $hosts, $params['timeout'], $params['read_timeout'], (bool) $params['persistent'] ?? false, $auth, ...\defined('Redis::SCAN_PREFIX') ? [$params['ssl'] ?? null] : []);
```
`(bool) ($params['persistent'] ?? false)`

maxbeckers committed:
Fix Integration tests skipped with Redis cluster

nicolas-grekas commented:
Please submit another PR for 5.4, I'll use this one when merging up.

nicolas-grekas committed:
bug #49422 [Cache][Messenger] fixed CallbackInterface support in async expiration handler (AdamKatzDev)

This PR was merged into the 5.4 branch.

Discussion
----------

[Cache][Messenger] fixed CallbackInterface support in async expiration handler

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

Adds support for CallbackInterface in EarlyExpirationHandler and thus fixes async cache recomputing for callables that implement CallbackInterface. Earlier similar errors were fixed in #31879.

Commits
-------

a8dea95119 [Messenger][Cache] fixed CallbackInterface support in async expiration handler

maxbeckers committed:
Update Redis version to 6.2.8

nicolas-grekas committed:
bug #49459 [Form] Skip password hashing on empty password (Seb33300)

This PR was squashed before being merged into the 6.2 branch.

Discussion
----------

[Form] Skip password hashing on empty password

| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

When using the new `hash_property_path` option to hash password submitted by forms, we should skip hashing if the submitted password is empty.

- Because empty passwords are not allowed and saving an empty password hash will prevent the user to login his account:
https://github.com/symfony/symfony/blob/8c19af2346ed6ed4021d3efc024e0af138e69794/src/Symfony/Component/Security/Http/EventListener/CheckCredentialsListener.php#L59-L62

- Because a common use case when creating a user profile form is to ignore the "new password" input if it's left blank.

Commits
-------

b4aa3ea79a [Form] Skip password hashing on empty password

nicolas-grekas committed:
bug #49421 [TwigBridge] do not drop embed label classes (xabbuh)

This PR was merged into the 5.4 branch.

Discussion
----------

[TwigBridge] do not drop embed label classes

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #45727
| License       | MIT
| Doc PR        |

Commits
-------

cd50683b25 do not drop embed label classes

nicolas-grekas committed:
bug #49405 [MonologBridge] FirePHPHandler::onKernelResponse throws PHP 8.1 deprecation when no user agent is set (juagarc4)

This PR was squashed before being merged into the 5.4 branch.

Discussion
----------

[MonologBridge] FirePHPHandler::onKernelResponse throws PHP 8.1 deprecation when no user agent is set

| Q             | A
| ------------- | ---
| Branch?       |  5.4
| Bug fix?      | yes/no
| New feature?  no
| Deprecations? no
| Tickets       | Fix #49392
| License       | MIT

- Add casting to the second parameter to pass always the right type.

**PHPstan before:**

 ------ --------------------------------------------------------------------------------------------------------
  Line   FirePHPHandler.php
 ------ --------------------------------------------------------------------------------------------------------
  27     Property Symfony\Bridge\Monolog\Handler\FirePHPHandler::$headers has no type specified.
  37     Method Symfony\Bridge\Monolog\Handler\FirePHPHandler::onKernelResponse() has no return type specified.
  **44     Parameter #2 $subject of function preg_match expects string, string|null given.**
  68     If condition is always true.
 ------ --------------------------------------------------------------------------------------------------------

**PHPstan after:**

 ------ --------------------------------------------------------------------------------------------------------
  Line   FirePHPHandler.php
 ------ --------------------------------------------------------------------------------------------------------
  27     Property Symfony\Bridge\Monolog\Handler\FirePHPHandler::$headers has no type specified.
  37     Method Symfony\Bridge\Monolog\Handler\FirePHPHandler::onKernelResponse() has no return type specified.
  68     If condition is always true.
 ------ --------------------------------------------------------------------------------------------------------

Commits
-------

4d84c46369 [MonologBridge] FirePHPHandler::onKernelResponse throws PHP 8.1 deprecation when no user agent is set

nicolas-grekas committed:
bug #49446 [SecurityBundle] Fix `Security::login()` on specific firewall (chalasr)

This PR was merged into the 6.2 branch.

Discussion
----------

[SecurityBundle] Fix `Security::login()` on specific firewall

| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | #49444
| License       | MIT
| Doc PR        | -

Commits
-------

f7043dbcb8 [SecurityBundle] Fix `Security::login()` on specific firewall

nicolas-grekas committed:
minor #49467 Update Redis version to 6.2.8 for integrationtests (maxbeckers)

This PR was merged into the 5.4 branch.

Discussion
----------

Update Redis version to 6.2.8 for integrationtests

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #49285
| License       | MIT
| Doc PR        | n/a

Related to #49449 to bring the redis versions all to the same version 6.2.8

/cc `@nicolas`-grekas

Commits
-------

a1696bf80e Update Redis version to 6.2.8

nicolas-grekas committed:
[Translation] Fix merge

nicolas-grekas committed:
Merge branch '5.4' into 6.2

* 5.4:
  [Translation] Fix merge
  [MonologBridge] FirePHPHandler::onKernelResponse throws PHP 8.1 deprecation when no user agent is set
  Update Redis version to 6.2.8
  [Contracts] Fix setting $container before calling parent::setContainer in ServiceSubscriberTrait
  [Messenger][Cache] fixed CallbackInterface support in async expiration handler
  do not drop embed label classes

nicolas-grekas committed:
Merge branch '6.2' into 6.3

* 6.2:
  [Translation] Fix merge
  [SecurityBundle] Fix `Security::login()` on specific firewall
  [MonologBridge] FirePHPHandler::onKernelResponse throws PHP 8.1 deprecation when no user agent is set
  [Form] Skip password hashing on empty password
  Update Redis version to 6.2.8
  [Contracts] Fix setting $container before calling parent::setContainer in ServiceSubscriberTrait
  [Messenger][Cache] fixed CallbackInterface support in async expiration handler
  do not drop embed label classes

nicolas-grekas commented:
Thank you @maxbeckers.
